### DISCUSSION TITLE
How to infer type from array item?

### DISCUSSION THREAD

**Discussion by cawa-93** [vimeo/psalm#10723] [Q&A]
I have the following code
```php
class AdminAssets
{

    public const ASSETS = [
        'js' => [
            'foo-script' => [
                'src' => 'file.js',
            ],
            'bar-script' => [
                'src' => 'file.js',
            ],
        ],
    ];

    public static function show(string $key)
    {
    }
}
```

How can I declare that method `show` can accept string union from `ASSETS['js']` keys (`'foo-script'|'bar-script'`)?

I'm [tired](https://psalm.dev/r/0f30dc4115) to use `@param key-of<ASSETS['js']> $key` but got error `[InvalidDocblock](https://psalm.dev/008) - 18:15 - Unrecognised template param 'js' in docblock for AdminAssets::show`

**Comment by TayfurYldz** [comment]
The part that Psalm supports directly is `key-of<CLASS_CONST>`; the unsupported part in the attempted annotation is indexing the constant inside the type expression with `['js']`.

Current Psalm documents/tests explicitly support forms such as:

```php
/** @param key-of<self::DAYS> $day */
```

and `key-of<Foo\Bar::ARRAY_CONST>`, but `ASSETS['js']` is not type-level array-offset syntax. That's why the parser ends up interpreting the bracketed expression incorrectly and reports the misleading template-param error.

If you need exactly the keys under `js`, the cleanest current representation is to expose that nested map as its own constant:

```php
final class AdminAssets
{
    public const JS = [
        'foo-script' => ['src' => 'file.js'],
        'bar-script' => ['src' => 'file.js'],
    ];

    public const ASSETS = [
        'js' => self::JS,
    ];

    /** @param key-of<self::JS> $key */
    public static function show(string $key): void
    {
    }
}
```

Now Psalm can derive `'foo-script'|'bar-script'` from the constant itself, without duplicating the string union in the docblock.

So `key-of` is the right utility; the limitation is that it operates on the array type/constant you give it and there isn't currently a supported `self::ASSETS['js']` projection syntax for selecting a nested offset inside that argument.
